core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -34,21 +46,182 @@ pub enum DiscoveryInitError {
+            "Binding handler for broadcast topic {}.",
+            OfferReceived::TOPIC
+        );
+
+        let _ = bus::bind_with_caller(
+            &offer_broadcast_address,
+            move |caller, msg: SendBroadcastMessage<OfferReceived>| {
+                let myself = myself.clone();
+                on_offer_received(myself, caller, msg.body().to_owned())
+            },
+        );
+
+        Ok(())
+    }
+}
+
+async fn broadcast_offer(myself: Arc<DiscoveryImpl>, offer: Offer) -> Result<(), DiscoveryError> {
+    log::info!("Broadcasting offer [{}] to the network.", offer.offer_id()?);
+
+    let msg = OfferReceived { offer };
+    let bcast_msg = SendBroadcastMessage::new(msg);
+
+    let _ = bus::service(net::local::BUS_ID).send(bcast_msg).await?;
+    Ok(())
 }
 
-/// Creates Discovery of specific type.
-pub trait DiscoveryFactory {
-    fn new(builder: DiscoveryBuilder) -> Result<Arc<dyn Discovery>, DiscoveryInitError>;
+async fn retrieve_offers(myself: Arc<DiscoveryImpl>) -> Result<Vec<Offer>, DiscoveryError> {
+    unimplemented!()
 }
 
+async fn on_offer_received(
+    myself: Arc<DiscoveryImpl>,
+    caller: String,
+    msg: OfferReceived,
+) -> Result<(), ()> {
+    let callback = myself.offer_received.clone();
+
+    let offer = msg.offer.clone();
+    let offer_id = offer.offer_id().unwrap_or("{Empty id}").to_string();
+    let provider_id = offer.provider_id().unwrap_or("{Empty id}").to_string();
+
+    log::info!(
+        "Received broadcasted Offer [{}] from provider [{}]. Sender: [{}].",
+        offer_id,
+        provider_id,
+        &caller,
+    );
+
+    match callback.call(caller, msg).await? {
+        PropagateOffer::True => {
+            log::info!("Propagating further Offer [{}].", offer_id,);
+
+            // TODO: Should we retry in case of fail?
+            if let Err(error) = broadcast_offer(myself, offer).await {
+                log::error!(
+                    "Error propagating further Offer [{}] from provider [{}].",
+                    offer_id,
+                    provider_id
+                );
+            }
+        }
+        PropagateOffer::False(reason) => {
+            log::info!(
+                "Not propagating Offer [{}] for reason: {}.",
+                offer_id,
+                reason
+            );
+        }
+    }
+    Ok(())
+}
+
+// =========================================== //
+// Discovery messages
+// =========================================== //
+
+#[derive(Serialize, Deserialize, Display)]
+pub enum StopPropagateReason {
+    #[display(fmt = "Offer already exists in database")]
+    AlreadyExists,
+    #[display(fmt = "Error adding offer: {}", "_0")]
+    Error(String),
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum PropagateOffer {
+    True,
+    False(StopPropagateReason),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OfferReceived {
+    pub offer: Offer,
+}
+
+impl RpcMessage for OfferReceived {
```

Comment:
it is _de facto_ standard to define `RpcMessage`s within `ya-core-model`

Reply:
Only for external Messages. No one is supposed to use this from outside of Market